metabase/metabase
# Update query builder data selector for datasets

This PR adds datasets support for GUI query data source picker (feature epic #18659)

![before-after-img](https://user-images.githubusercontent.com/17258145/139724310-c0e7fba3-7a2a-4a13-831f-a0d200e49272.png)

When people have at least one dataset in Metabase, we're going to display a new first step. Otherwise, the picker should remain the same. About the "tabs":

* **Saved Questions** works the same way it does now
* **Raw Data** contains databases, schemas, and tables (the one you can see when using the picker now). The only difference here is that Raw Data doesn't list the "Saved Questions" option as they have their own tab
* **Datasets** tab looks the same way as the saved question picker, but only lists datasets

## Implementation Details

The goal is to have a minimal version of datasets on `master` ASAP, so we can play around with the whole experience and introduce changes early. There are some trade-offs I had to address, and the main principles in the decisions are speed and YAGNI in case we want to change something about the requirements.

### Reusing saved questions picker

Datasets picker is expected to look and work the same way as the saved question picker, but only display datasets. So we should either extend the existing saved questions picker to support datasets too (datasets are saved questions under the hood, so their APIs are almost identical) or extract some common code and create two different components. This PR _extends_ the saved questions picker, so the data selector uses _just one_ component for both datasets and saved questions.

### Encoding extra parameters in schema ID

As mentioned above, this PR actually uses the saved questions picker for datasets too. When a user selects a collection from the picker, [`SavedQuestionList` requests a list of questions from the backend using the Schema entity loader](https://github.com/metabase/metabase/blob/5fcf4a3aa216298ec12a2d380668df0ab33d9e16/frontend/src/metabase/query_builder/components/saved-question-picker/SavedQuestionList.jsx#L50).

The trick here is that the BE will only return saved questions (not datasets) as expected. And we need to use another API endpoint to get datasets from a particular collection. So we need a way to let the Schema entity loader know when we need questions and when we need datasets. 

This PR extends the schema ID string to also contain some extra JSON data. Schema IDs currently look like `14:Application` where `14` is the database's ID and `Application` is a name of a schema. For saved questions, schema IDs look like `-1337:Marketing` where `-1337` is a "virtual" database with saved questions and `Marketing` is a name of a collection. This PR extends the format, so we can now also have something like `-1337:Marketing:{"isDatasets":true}`. Schema entity later [uses this parameter](https://github.com/metabase/metabase/blob/139ee2cea6460cc0a4183ddc59d86de9922d37b2/frontend/src/metabase/entities/schemas.js#L41) to decide which endpoint it should use.

### Schema file reorg

There is a [`schema` file in FE source code root](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/schema.js). It has entity schema definitions and some extra helpers for encoding and decoding schema ID, etc. I needed to use `generateSchemaId` in `metabase/lib/saved-questions` ([here](https://github.com/metabase/metabase/blob/139ee2cea6460cc0a4183ddc59d86de9922d37b2/frontend/src/metabase/lib/saved-questions/saved-questions.js#L19)), but the FE broke because some intertwined imports order if I just imported it. So in this PR, the schema utils are moved to [`metabase/lib/schema`](https://github.com/metabase/metabase/blob/139ee2cea6460cc0a4183ddc59d86de9922d37b2/frontend/src/metabase/lib/schema/schema.js). Entity schemas are still kept in `metabase/schema`

## To Verify

1. Archive any datasets if you have any
2. Ask a question > Simple / Custom Question
3. Make sure the data source picker looks and works the same way it does on `master`
4. Turn any question into a dataset (you can follow the steps from #18702)
5. Ask a question > Simple / Custom Question
6. The picker should have a new look now. You should see three options: Datasets, Raw Data, Saved Questions. Don't click anything for now

**Common Search**

Try to find a dataset, a table, and a saved question by their names. The search shouldn't skip any of these models

**Raw Data Tab**

1. Click the "Raw Data" tab
2. You should not see "Saved Questions" DB anywhere in the list
3. Try to find a dataset, a table, and a saved question by their names. While being on the "Raw Data" tab, the search should be scoped to tables. So you shouldn't see any datasets and saved questions in the results
4. Make sure the picker works the same way it does on `master` (expanding DB schemas, selecting things, etc.)
5. Click the back button, you should appear on the list with "Datasets", "Raw Data" and "Saved Questions"

**Datasets Tab**

1. Click the "Datasets" tab
2. Try to find a dataset, a table, and a saved question by their names. While being on the "Datasets" tab, the search should be scoped to datasets. So you shouldn't see any tables and saved questions in the results
3. Click on a few collections. If a collection has datasets, they should be shown on the right side
4. Make sure you don't see saved questions in the collections

**Saved Questions Tab**

1. Click the "Saved Questions" tab
2. Try to find a dataset, a table, and a saved question by their names. While being on the "Saved Questions" tab, the search should be scoped to saved questions only. So you shouldn't see any datasets and tables in the results
3. Make sure the picker works the same way it does on `master` (clicking collections, displaying content, etc.)
4. Make sure you don't see a dataset in the collection you saved it to
5. Click the back button, you should appear on the list with "Datasets", "Raw Data" and "Saved Questions"

**Ask a question!**

Create and save a question based on a dataset

## Demo

https://user-images.githubusercontent.com/17258145/139733530-ea3182f6-a820-407d-b8db-3490859fc061.mov
